pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,1016 @@
+// This file is part of pallet-dca.
+
+// Copyright (C) 2020-2023  Intergalactic, Limited (GIB).
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # DCA pallet
+//!
+//! ## Overview
+//!
+//! A dollar-cost averaging pallet that enables users to perform repeating orders.
+//!
+//! When an order is submitted, it will reserve the total amount (budget) specified by the user.
+//!
+//! A named reserve is allocated for the reserved amount of all DCA held by each user.
+//!
+//! The DCA plan is executed as long as there is remaining balance in the budget.
+//!
+//! If a trade fails due to specific errors whitelisted in the pallet config,
+//! then retry happens up to the maximum number of retries specified also as config.
+//! Once the max number of retries reached, the order is terminated permanently.
+//!
+//! If a trade fails due to other kind of errors, the order is terminated permanently without any retry logic.
+//!
+//! Orders are executed on block initialize and they are sorted based on randomness derived from relay chain block number.
+//! Therefore they cannot be front-ran in the block they are executed.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::MaxEncodedLen;
+use frame_support::{
+	ensure,
+	pallet_prelude::*,
+	traits::{Get, Len},
+	transactional,
+	weights::WeightToFee as FrameSupportWeight,
+};
+use frame_system::{ensure_signed, pallet_prelude::OriginFor, Origin};
+use hydradx_traits::pools::SpotPriceProvider;
+use hydradx_traits::{OraclePeriod, PriceOracle};
+use orml_traits::arithmetic::CheckedAdd;
+use orml_traits::MultiCurrency;
+use orml_traits::NamedMultiReservableCurrency;
+use rand::rngs::StdRng;
+use rand::{Rng, SeedableRng};
+use scale_info::TypeInfo;
+use sp_runtime::traits::CheckedMul;
+use sp_runtime::{
+	traits::{BlockNumberProvider, Saturating},
+	ArithmeticError, BoundedVec, DispatchError, FixedPointNumber, FixedU128, Permill,
+};
+use sp_std::vec::Vec;
+use sp_std::{
+	cmp::{max, min},
+	vec,
+};
+#[cfg(test)]
+mod tests;
+
+#[cfg(any(feature = "runtime-benchmarks", test))]
+mod benchmarks;
+
+pub mod types;
+pub mod weights;
+
+use weights::WeightInfo;
+
+// Re-export pallet items so that they can be accessed from the crate namespace.
+pub use pallet::*;
+
+use crate::types::*;
+
+type BlockNumberFor<T> = <T as frame_system::Config>::BlockNumber;
+
+const MAX_NUMBER_OF_RETRY_FOR_PLANNING: u32 = 5;
```

Comment:
this is not about how many retries we do in case of error.

It is about how many times we try to look for a free block we can use to plan the schedule

You remember, we try to search for free blocks with <block> + <delay> wher delay is 1 - 2 - 4 - 8 - 16. 

If the x+16 block is still full, we terminate.

I think the naming is not good for this, so I will rename to `RETRY_TO_SEARCH_FOR_FREE_BLOCK` to make it more clear.

Reply:
ok, then mby moving it to function or close to it would be better, it is hard to remember property that will come in play in 700 LOC :)

Reply:
I have seen that consts are always defined at the beginning of the `lib.rs` pallet files, across most of our pallets, therefore I think it is a good place for it.